storm-core/src/jvm/org/apache/storm/daemon/supervisor/Container.java

```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * Is a process alive and running?
+     * @param pid the PID of the running process
+     * @param user the user that is expected to own that process
+     * @return true if it is, else false
+     * @throws IOException on any error
+     */
+    protected boolean isProcessAlive(long pid, String user) throws IOException {
+        if (Utils.IS_ON_WINDOWS) {
+            return isWindowsProcessAlive(pid, user);
+        }
+        return isPosixProcessAlive(pid, user);
+    }
+    
+    private boolean isWindowsProcessAlive(long pid, String user) throws IOException {
+        boolean ret = false;
+        ProcessBuilder pb = new ProcessBuilder("tasklist", "/nh", "/fi", "pid eq"+pid);
+        pb.redirectError(Redirect.INHERIT);
+        Process p = pb.start();
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
+            if (in.readLine() != null) {
+                ret = true;
+            }
+        }
+        return ret;
+    }
+    
+    private boolean isPosixProcessAlive(long pid, String user) throws IOException {
+        boolean ret = false;
+        ProcessBuilder pb = new ProcessBuilder("ps", "-o", "user", "-p", String.valueOf(pid));
+        pb.redirectError(Redirect.INHERIT);
+        Process p = pb.start();
+        try (BufferedReader in = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
+            String first = in.readLine();
+            assert("USER".equals(first));
+            String processUser;
+            while ((processUser = in.readLine()) != null) {
+                if (user.equals(processUser)) {
+                    ret = true;
+                    break;
+                } else {
+                    LOG.info("Found {} running as {}, but expected it to be {}", pid, processUser, user);
+                }
+            }
+        }
+        return ret;
+    }
+    
+    @Override
+    public boolean areAllProcessesDead() throws IOException {
+        Set<Long> pids = getAllPids();
+        String user = getWorkerUser();
+        
+        boolean allDead = true;
+        for (Long pid: pids) {
+            if (!isProcessAlive(pid, user)) {
+                LOG.warn("{}: PID {} is dead", _workerId, pid);
+            } else {
+                allDead = false;
+                break;
+            }
+        }
+        return allDead;
+    }
+
+    @Override
+    public void cleanUp() throws IOException {
+        cleanUpForRestart();
+    }
+
+    /**
+     * Setup the container to run.  By default this creates the needed directories/links in the
+     * local file system
+     * PREREQUISITE: All needed blobs and topology, jars/configs have been downloaded and
+     * placed in the appropriate locations
+     * @throws IOException on any error
+     */
+    protected void setup() throws IOException {
+        if (_port <= 0) {
+            throw new IllegalStateException("Cannot setup a container recovered with just a worker id");
```

Comment:
Could we clarify this a bit? I wasn't exactly sure what this meant with this wording.